### PR TITLE
[WebProfiler] added support for window.fetch calls in ajax section

### PR DESCRIPTION
**Pull request by ivoba**
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17444 |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

This adds support for window.fetch calls to the Ajax section of the WebProfiler toolbar.

Credits to @tbopec for  implementation :)

### CODE REVIEW

**Comment by fabpot** [comment]
@javiereguiluz Can you validate this one?

**Comment by fabpot** [comment]
@robfrawley Can you test it?

**Comment by nicolas-grekas** [comment]
Could anyone please test this one?

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 1]
```diff
@@ -236,6 +236,31 @@
...
+                            start: new Date()
+                        };
+                        requestStack.push(stackElement);
+                        promise.then(function (r) {
+                            stackElement.duration = new Date() - stackElement.start;
+                            stackElement.loading = false;
+                            stackElement.error = r.status < 200 || r.status >= 400;
+                            stackElement.statusCode = r.status;
+                            stackElement.profile = r.headers.get('x-debug-token');
+                            stackElement.profilerUrl = r.headers.get('x-debug-token-link');
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: Is this compatible with CORS restriction or will it log a security warning ?

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 2]
```diff
@@ -236,6 +236,31 @@
 
         {% if excluded_ajax_paths is defined %}
             if (window.XMLHttpRequest && XMLHttpRequest.prototype.addEventListener) {
+                if (window.fetch) {
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: this does not need to be inside the check for ``window.XMLHttpRequest``

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 3]
```diff
@@ -236,6 +236,31 @@
...
+                    var oldFetch = window.fetch;
+                    window.fetch = function () {
+                        var promise = oldFetch.apply(null, arguments);
+                        var stackElement = {
+                            loading: true,
+                            error: false,
+                            url: arguments[0],
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: you need to match the URL against the excluded ajax paths before adding it in the request stack

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 4]
```diff
@@ -236,6 +236,31 @@
...
+                            method: arguments[1].method,
+                            start: new Date()
+                        };
+                        requestStack.push(stackElement);
+                        promise.then(function (r) {
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: what if the promise gets rejected ? The stack element should not stay in loading state in this case

**Comment by stof** [comment]
Status: needs work

**Comment by ivoba** [comment]
Applied fixes for @stof `s comments.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<script{% if csp_script_nonce is defined and csp_script_nonce %} nonce={{ csp_script_nonce }}{% endif %}>/*<![CDATA[*/
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: This remove must be reverted. You are breaking the CSP compatibility

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 6]
```diff
@@ -22,204 +22,204 @@
 
         var noop = function() {},
 
-            collectionToArray = function (collection) {
-                var length = collection.length || 0,
-                    results = new Array(length);
+                collectionToArray = function (collection) {
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: please revert all these indentation changes

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 7]
```diff
@@ -235,77 +235,94 @@
...
-                    };
-                }
-                var proxied = XMLHttpRequest.prototype.open;
 
-                XMLHttpRequest.prototype.open = function(method, url, async, user, pass) {
-                    var self = this;
+        if (window.fetch) {
+            var oldFetch = window.fetch;
+            window.fetch = function () {
+                if (!arguments[0].match(new RegExp({{ excluded_ajax_paths|json_encode|raw }}))) {
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: applying the original fetch (and returning the promise) must still be made even when the URL is excluded from the profiling, otherwise you are breaking ``window.fetch``

**Comment by stof** [comment]
I will review again when the indentation changes are reverted. They make it very hard to see what actually changed (and if I ignored the whitespace changes, I cannot comment on the diff anymore, which is not convenient)

**Comment by ivoba** [comment]
oops how did that happen?
fix is coming...

**Comment by ivoba** [comment]
@stof done, sorryy

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 8]
```diff
@@ -235,6 +235,46 @@
...
+
+            if (window.fetch) {
+                var oldFetch = window.fetch;
+                window.fetch = function () {
+                    if (!arguments[0].match(new RegExp({{ excluded_ajax_paths|json_encode|raw }}))) {
+
+                        var promise = oldFetch.apply(null, arguments);
+
+                        var method = 'GET';
+                        if(arguments[1].method !== undefined){
```
Comment (tsm91) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: @ivoba The `init` config object is optional (`Promise<Response> fetch(input[, init]);`) therefore this can throw an error. First check if the 2nd argument even exists `arguments[1] && arguments[1].method !== undefined` .

Comment (ivoba) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: @tsm91 correct. i fixed this.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 1 above)
Comment (fabpot) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: @romainneutron any ideas?

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 9]
```diff
@@ -235,6 +235,47 @@
         }
 
         {% if excluded_ajax_paths is defined %}
+
```
Comment (fabpot) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: extra blank line here

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 10]
```diff
@@ -235,6 +235,47 @@
...
+            if (window.fetch) {
+                var oldFetch = window.fetch;
+                window.fetch = function () {
+                    if (!arguments[0].match(new RegExp({{ excluded_ajax_paths|json_encode|raw }}))) {
+
```
Comment (fabpot) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: extra blank line here

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 11]
```diff
@@ -235,6 +235,47 @@
...
+                        var promise = oldFetch.apply(null, arguments);
+
+                        var method = 'GET';
+
+                        if(arguments[1] && arguments[1].method !== undefined){
```
Comment (fabpot) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: missing space after `if` and before `{`

**Comment by Gadgetdude** [comment]
Really looking forward to this fix/feature. Is there any way I can assist?

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 12]
```diff
@@ -235,6 +235,45 @@
         }
 
         {% if excluded_ajax_paths is defined %}
+
+            if (window.fetch) {
+                var oldFetch = window.fetch;
+                window.fetch = function () {
+                    if (!arguments[0].match(new RegExp({{ excluded_ajax_paths|json_encode|raw }}))) {
+                        var promise = oldFetch.apply(null, arguments);
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: this call must be done outside the URL profiling blacklist check (and the return at the end too). Otherwise, you break fetch entirely for URLs blacklisted by the AJAX profiling

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 12 above)
Comment (ivoba) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: @stof fixed

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 1 above)
Comment (romainneutron) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: I'm gonna try this PR

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 13]
```diff
@@ -235,6 +235,44 @@
         }
 
         {% if excluded_ajax_paths is defined %}
+
+            if (window.fetch) {
+                var oldFetch = window.fetch;
+                window.fetch = function () {
+                    var promise = oldFetch.apply(null, arguments);
```
Comment (romainneutron) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: `oldFetch.apply(window, arguments);`

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 14]
```diff
@@ -235,6 +235,44 @@
...
+                        requestStack.push(stackElement);
+                        promise.then(function (r) {
+                            stackElement.duration = new Date() - stackElement.start;
+                            stackElement.loading = false;
+                            stackElement.error = r.status < 200 || r.status >= 400;
+                            stackElement.statusCode = r.status;
+                            stackElement.profile = r.headers.get('x-debug-token');
+                            stackElement.profilerUrl = r.headers.get('x-debug-token-link');
+                            Sfjs.renderAjaxRequests();
+                        }).catch(function(err) {
```
Comment (romainneutron) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: it's better to use the second argument of `then` instead of using `catch`: if an error is throws by `Sfjs.renderAjaxRequests`, the request would be tagged as erroneous whereas the error came from the code executed after that.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 15]
```diff
@@ -235,6 +235,44 @@
...
+                            stackElement.loading = false;
+                            stackElement.error = true;
+                        });
```
Comment (romainneutron) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: what about adding the error content as a property and display this error in the toolbar?

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 13 above)
Comment (romainneutron) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: Works perfectly, no security error, even in `cors` mode

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 16]
```diff
@@ -235,6 +235,44 @@
...
+                            stackElement.duration = new Date() - stackElement.start;
+                            stackElement.loading = false;
+                            stackElement.error = r.status < 200 || r.status >= 400;
+                            stackElement.statusCode = r.status;
+                            stackElement.profile = r.headers.get('x-debug-token');
+                            stackElement.profilerUrl = r.headers.get('x-debug-token-link');
+                            Sfjs.renderAjaxRequests();
+                        }).catch(function(err) {
+                            stackElement.loading = false;
+                            stackElement.error = true;
```
Comment (romainneutron) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: what about adding a `type` so we could make distinction between `ajax (xhr)` and `fetch` in the toolbar. As they do not behave the same way, it may be useful, WDYT?

**Review by romainneutron — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 17]
```diff
@@ -235,6 +235,44 @@
         }
 
         {% if excluded_ajax_paths is defined %}
+
+            if (window.fetch) {
+                var oldFetch = window.fetch;
+                window.fetch = function () {
```
Comment (romainneutron) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: what about using `fetch` signature (`fetch(url, options):promise`) instead of relying on arguments? I think it's easier to read

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 16 above)
Comment (ivoba) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: yeah was thinking about that too, but was bit afraid of the extra effort. i see what i can do.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 13 above)
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: even better: ``.apply(this, arguments``, keeping the same context

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig [hunk 18]
```diff
@@ -235,6 +235,44 @@
         }
 
         {% if excluded_ajax_paths is defined %}
+
+            if (window.fetch) {
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: should we try to detect native fetch here ? If someone uses a fetch polyfill based on XHR (e.g. the polyfill maintained by github), we would register the request twice otherwise (at the fetch level and at the xhr level)

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 16 above)
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: well, the extra work is mainly about figuring out the display. Adding the type property is not hard (it should just be done in both place initializing a stackElement)

**Comment by HeahDude** [comment]
@ivoba Do you need help with the templates so you can finish this feature for 3.2? :)

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 18 above)
Comment (ivoba) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: how would you detect this? just check if the browser supports fetch? this needs to happen before any ployfill is loaded, im not sure how to do this.
also as the profiler is a dev tool, how many people develop on outdated browsers?

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 13 above)
Comment (ivoba) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: fixed using @stof 's approach

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 14 above)
Comment (ivoba) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: fixed

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 16 above)
Comment (ivoba) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: i added type to stackElement and to display

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 17 above)
Comment (ivoba) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: absolutly, fixed

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 15 above)
Comment (ivoba) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: its a good idea, but maybe we can put this in another PR since its also not implemented for XHR, afaik, and it would need some more adjustments in how we render the table.

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig (hunk 18 above)
Comment (ivoba) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig]: ok i see, i added a check on polyfill `if (window.fetch && window.fetch.polyfill === undefined)`

File: src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar.css.twig [hunk 19]
```diff
@@ -334,7 +334,7 @@
     padding: 4px;
 }
 .sf-ajax-request-url {
-    max-width: 300px;
+    max-width: 250px;
```
Comment (stof) [inline src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar.css.twig]: why changing this ?

**Review by stof — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @ivoba.
